### ISSUE TITLE
Compile times

### ISSUE THREAD

**Issue by Timmmm** [JonathanWoollett-Light/ux2#19]
Any idea why compile times are *so* bad? I was hoping to use this library but 45m is way too much. Did the original uX crate take that long too?

**Comment by Timmmm** [comment]
> Did the original uX crate take that long too?

I tested it. Takes 30 seconds on my 10 year old desktop. I tried `ux2` and it didn't complete in a minute and memory usage went over 10GB before I had to kill it. That is a crazy disparity. I don't think I would consider using this crate while that is the case but I'd like to because of all the other improvements.

Do you have any idea what causes it?

**Comment by GnomedDev** [comment]
I believe it's the enum based approach, rustc is pretty bad at compiling enums with insane numbers of variants. If possible, you should probably avoid the 128 feature.

**Comment by JonathanWoollett-Light** [comment]
ux2 generates all the combinations of arithmetic operations e.g. `i1 + i1 = i2`, `i1 + i2 = i3` etc. this results in millions of lines of code at 64 and 128. When compiled all these lines get optimized out, but it really stresses the compiler.

I will be looking at optimising this.

**Comment by GnomedDev** [comment]
Could these arithmetic operations get their own feature, instead of feature gating the whole type?

**Comment by JonathanWoollett-Light** [comment]
> Could these arithmetic operations get their own feature, instead of feature gating the whole type?

Yes it may make sense to put the cross-type arithmetic operations (e.g. `i1 + i2`) behind a feature flag.

**Comment by JonathanWoollett-Light** [comment]
I have not given myself enough time to maintain this project as well as I'd like, apologies for this.

I have posted https://www.reddit.com/r/rust/comments/17pljet/how_can_i_speed_up_compilation_with_procmacros/ to look into the topic.

The approach I'd like to take is improving compile time and retaining the benefits of the current approach rather than discarding these benefits to recoup compile time.
